client/post-editor/post-editor.jsx

```diff
@@ -1013,17 +1014,268 @@ export const PostEditor = React.createClass( {
+		if ( isCursorEndInTag ) {
+			htmlModeCursorEndPosition = isCursorEndInTag.gtPos;
+		}
+
+		let mode = 'single';
+		if ( htmlModeCursorStartPosition !== htmlModeCursorEndPosition ) {
+			mode = 'range';
+		}
+
+		const bookMarkStart = '<span ' +
+			'data-mce-type="bookmark"	' +
+			'id="mce_SELREST_start" ' +
+			'data-mce-style="overflow:hidden;line-height:0px" ' +
+			'style="overflow:hidden;line-height:0px"' +
+			'>' +
+			'&#65279;' +
+			'</span>';
+
+		let bookMarkEnd = null;
+		if ( mode === 'range' ) {
+			bookMarkEnd = '<span ' +
+				'data-mce-type="bookmark"	' +
+				'id="mce_SELREST_end" ' +
+				'data-mce-style="overflow:hidden;line-height:0px" ' +
+				'style="overflow:hidden;line-height:0px"' +
+				'>' +
+				'&#65279;' +
+				'</span>';
+
+			// // TODO detect if you're not in a tag to not break things
+			textArea.value = [
+				textArea.value.slice( 0, htmlModeCursorStartPosition ),
+				bookMarkStart,
+				textArea.value.slice( htmlModeCursorStartPosition, htmlModeCursorEndPosition ),
+				bookMarkEnd,
+				textArea.value.slice( htmlModeCursorEndPosition ),
+			].join( '' );
+		} else {
+			textArea.value = [
+				textArea.value.slice( 0, htmlModeCursorStartPosition ),
+				bookMarkStart,
+				textArea.value.slice( htmlModeCursorStartPosition )
+			].join( '' );
+		}
+
+		this.editor.onTextAreaChange( { target: { value: textArea.value } } );
+	},
+
+	focusHTMLBookmarkInVisualEditor: function( ed ) {
+		const startNode = ed.target.getDoc().getElementById( 'mce_SELREST_start' );
+		const endNode = ed.target.getDoc().getElementById( 'mce_SELREST_end' );
+
+		if ( ! startNode ) {
+			return;
+		}
+
+		ed.target.focus();
+
+		if ( ! endNode ) {
+			ed.target.selection.select( startNode );
+		} else {
+			const sel = document.createRange();
+			sel.setStart( startNode, 0 );
+			sel.setEnd( endNode, 0 );
+
+			ed.target.selection.setRng( sel );
+			endNode.parentNode.removeChild( endNode );
+		}
+
+		// TODO: Scroll to cursor position?
+		startNode.parentNode.removeChild( startNode );
+
+		ed.target.off( 'SetContent', this.focusHTMLBookmarkInVisualEditor );
+	},
+
+	/**
+	 * Finds the current selection position in the Visual editor.
+	 *
+	 * It uses some black magic raw JS magic. Not for the faint-hearted.
+	 *
+	 * @param {Object} editor The editor where we must find the selection
+	 * @returns {null | Object} The selection range position in the editor
+	 */
+	findBookmarkedPosition: function( editor ) {
+		// Get the TinyMCE `window` reference, since we need to access the raw selection.
+		const TinyMCEWIndow = editor.getWin();
+
+		const selection = TinyMCEWIndow.getSelection();
+
+		if ( selection.rangeCount <= 0 ) {
+			// no selection, no need to continue.
+			return;
+		}
+
+		/**
+		 * The ID is used to avoid replacing user generated content, that may coincide with the
+		 * format specified below.
+		 * @type {string}
+		 */
+		const selectionID = 'SELRES_' + Math.random();
```

Comment:
Would `uuid` be better here (I don't know)

Reply:
Code updated, check [this comment](https://github.com/Automattic/wp-calypso/pull/17481#discussion_r135399392).